server/id/reusable_id_impl.go

```diff
@@ -0,0 +1,53 @@
+package id
+
+import (
+	"context"
+	"sort"
+	"sync"
+)
+
+type ReusableAllocatorImpl struct {
+	// RWMutex is used to protect following fields.
+	lock sync.Mutex
+
+	minID    uint64
+	existIDs []uint64
```

Comment:
Actually, the most suitable struct here is an ordered container, such as an ordered map. However, there is no such containers in the standard library.

I prefer to encapsulate a structure called `OrderedList` here with the methods:
```
type OrderedList = []uint64

func (l OrderedList) IndexOfValue(v uint64) int
func (l OrderedList) InsertAt(v uint64, at int)
func (l OrderedList) Insert(v int64)
```

After that, the code will be more concise and elegant.